Remake bundle in docker

gkurz opened the pull request:
When bumping OSC versions or any other change that affects the final CSV, we need to run `make bundle`. This is a manual action that is error prone. Also `make bundle` always do extra changes that we don't want to merge (e.g. changing image digests into tags or reverting wanted changes in bundle related files).

Add all the logic to call `make bundle` from a container :
- add `controller-gen` as done in the operator build for hermetic builds
- add `kustomize` as a dep in go.mod and do like `controller-gen` for hermetic builds
- add  `operator-sdk` using hermeto's generic fetcher

The OSC operator image is now stored in the `config/manager/manager.yaml` file for reference. It is monitored by konflux and passed as `IMG` to `make bundle`. This requires the `yq` command. Use the same workaround as `controller-gen` to honor hermetic builds in konflux.

gkurz commented:
/retest

gkurz committed:
Prefetch go deps for OSC operator bundle builds

This will be needed to be able to run `make bundle` hermetically.

Signed-off-by: Greg Kurz <groug@kaod.org>

gkurz commented:
/retest

gkurz commented:
/retest

gkurz commented:
/ok-to-test

littlejawa commented on bundle.Dockerfile:
```diff
@@ -1,3 +1,41 @@
+COPY controller-gen bin/
+
+# get the version of controller-gen in an env variable for reusing
+RUN echo "export CONTROLLER_TOOLS_VERSION=$(grep controller-tools go.mod | awk '{print $2}')" > controller-tools-ver.env
```
```suggestion
RUN echo "export CONTROLLER_TOOLS_VERSION=$(grep controller-tools go.mod | awk '{print $2}')" > bundle-build.env
```

littlejawa commented on bundle.Dockerfile:
```diff
@@ -1,3 +1,41 @@
+COPY yq bin/
+
+# save the current version of the operator image in an env variable for reusing
+RUN echo "export IMG=$(./bin/yq .spec.install.spec.deployments[0].spec.template.spec.containers[0].image bundle/manifests/sandboxed-containers-operator.clusterserviceversion.yaml)" > operator-image.env
```
```suggestion
RUN echo "export IMG=$(./bin/yq .spec.install.spec.deployments[0].spec.template.spec.containers[0].image bundle/manifests/sandboxed-containers-operator.clusterserviceversion.yaml)" >> bundle-build.env
```

littlejawa commented on bundle.Dockerfile:
```diff
@@ -1,3 +1,41 @@
+RUN echo "export IMG=$(./bin/yq .spec.install.spec.deployments[0].spec.template.spec.containers[0].image bundle/manifests/sandboxed-containers-operator.clusterserviceversion.yaml)" > operator-image.env
+
+# make sure 'make bundle' uses the right version of controller-gen and preserves the operator image
+RUN . ./controller-tools-ver.env && . ./operator-image.env && make bundle
```
```suggestion
RUN . ./bundle-build.env && make bundle
```

littlejawa reviewed (commented):
I think we can use a single env file for the parameters we give to "make"

gkurz committed:
Remake bundle in docker

When bumping OSC versions or any other change that affects the final
CSV, we need to run `make bundle`. This is a manual action that is
error prone. Also `make bundle` always do extra changes that we don't
want to merge (e.g. changing image digests into tags or reverting
wanted changes in bundle related files).

Add all the logic to call `make bundle` from a container.

Signed-off-by: Greg Kurz <groug@kaod.org>

gkurz committed:
Add yq wrapper

This will be needed to be able to run `yq` hermetically.

Signed-off-by: Greg Kurz <groug@kaod.org>

gkurz committed:
Prefetch operator-sdk

This will be needed to be able to run `operator-sdk` hermetically.

Signed-off-by: Greg Kurz <groug@kaod.org>

gkurz commented:
> I think we can use a single env file for the parameters we give to "make"

Heh I did that already locally :smile:

gkurz committed:
Add kustomize wrapper

This will be needed to be able to run `make bundle` hermetically.

Bump kustomize to v5.6.0 to fix deps.

Signed-off-by: Greg Kurz <groug@kaod.org>

gkurz commented:
/retest

littlejawa reviewed (approved):
/lgtm
Thanks @gkurz !

gkurz commented:
I need to check the resulting bundle is as expected.

/hold

gkurz committed:
Remove bundle from git

This isn't needed anymore since we run `make bundle` in konflux.
Now treat the `bundle` directory as a regular generated artifact
that should be removed by `make clean`.

Signed-off-by: Greg Kurz <groug@kaod.org>

gkurz committed:
Add operator image to manager.yaml

The current operator image is only stored in the CSV within the `bundle`
directory that will be removed from the git repo soon. Record the image
as a related image in manager.yaml and have the tekton pipelines to
monitor it for updates. This will be used to build the bundle with
konflux.

Signed-off-by: Greg Kurz <groug@kaod.org>

gkurz commented:
Too complex.
